### ISSUE TITLE
Managing options and secrets in .NET Core and Docker

### ISSUE THREAD

**Issue by matjanos** [matjanos/matjanowski-blog#3]

**Comment by matjanos** [comment]
test 1 
`additional `

**Comment by matjanos** [comment]
test 123123

**Comment by matjanos** [comment]
code example:

```csharp

public void Test(){
return;
```

**Comment by PeterMilovcik** [comment]
I'm not able to use `builder.AddUserSecrets<Startup>();` - cannot resolve symbol AddUserSecrets.
Which NuGet package do I need for this?

**Comment by matjanos** [comment]
hi @PeterMilovcik ,
It is [here](https://www.nuget.org/packages/Microsoft.Extensions.Configuration.UserSecrets/)

**Comment by jpeckham** [comment]
I didn't have to add that manually... is it because i'm using WebApplication.CreateBuilder ?

```
var builder = WebApplication.CreateBuilder(args);
```

**Comment by Umut-Numanoglu-Docplanner** [comment]
Is there any more roboust way of doing it after nearly 3 years? 
My problem is that the USER_SECRETS_ID is different for everybody, so everybody needs to change the docker-compose file for themselves.
Also there are extra steps for securing the credentials on each machine, sending them to each developer and making sure to update it on each machine when something changes sound very counterproductive to me.
Before C# I was using ansible for managing our project, which offers an encrypted vault, which can be used with an agent. Maybe there is something similar from Microsoft? Or should I look into 3rd party tools like Harshicorp Vault?

**Comment by darcythomas-coop** [comment]
> Is there any more roboust way of doing it after nearly 3 years? My problem is that the USER_SECRETS_ID is different for everybody, so everybody needs to change the docker-compose file for themselves.

A couple of options:

 1. Have the same USER_SECRETS_ID for all developers; It *should* be unique per project/solution, **not** unique per developer
 2. create an `.env` file per developer with their USER_SECRETS_ID; Make sure you add  `.env` to the `.gitignore` file

**Comment by matjanos** [comment]
> I didn't have to add that manually... is it because i'm using WebApplication.CreateBuilder ?
> 
> ```
> var builder = WebApplication.CreateBuilder(args);
> ```
@jpeckham, exactly. In newer .NET versions (starting with .NET 6’s minimal hosting model), you typically use WebApplication.CreateBuilder(args), which already sets up configuration (including user secrets) by default. So you don’t need the old-school ConfigurationBuilder code in your Startup. The framework does that wiring for you under the hood.

**Comment by matjanos** [comment]
> Is there any more roboust way of doing it after nearly 3 years? My problem is that the USER_SECRETS_ID is different for everybody, so everybody needs to change the docker-compose file for themselves. Also there are extra steps for securing the credentials on each machine, sending them to each developer and making sure to update it on each machine when something changes sound very counterproductive to me. Before C# I was using ansible for managing our project, which offers an encrypted vault, which can be used with an agent. Maybe there is something similar from Microsoft? Or should I look into 3rd party tools like Harshicorp Vault?

@Umut-Numanoglu-Docplanner, Yeah, user secrets are really meant for local dev only, so they can get pretty messy if you try scaling them beyond that. Typically, you’d use something like [Azure Key Vault](https://learn.microsoft.com/azure/key-vault/general/overview) for storing sensitive info if you’re in the Microsoft ecosystem—this offloads that secret management headache and syncs well with Azure services.

For an on-prem or multi-cloud setup, HashiCorp Vault is a solid pick. It’s widely used, integrates with many platforms, and can handle both local dev secrets and production secrets in a more secure and centralized way. If you’re already comfortable with Ansible’s approach, adopting Vault or a similar tool might feel natural and definitely more robust than everyone juggling their own USER_SECRETS_ID.
